### PR TITLE
feat(TDP-5966): add country converter

### PR DESCRIPTION
**Pull request by tlnd-cgall**
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDP-5966

**Please check if the PR fulfills these requirements**
- [x] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] The code coverage on new code is > 75 % for backend and > 95% for frontend
- [ ] The new code does not introduce new technical issues (sonar / eslint)
- [x] Functional tests have been performed
- [ ] Docker configuration files for config-std or config-cloud profiles are impacted

**Please check the browsers you've tested on**
- [ ] Chrome, Firefox, Safari, Edge, IE11
- [ ] No, that's bad, this PR should not be merged !
- [ ] No, and no need to (backend changes only)

### CODE REVIEW

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+    protected static final String COUNTRY_NUMBER = "country_number";
+
+    private static final boolean CREATE_NEW_COLUMN_DEFAULT = true;
+
+    private static final String NEW_COLUMN_SEPARATOR = "_in_";
+
+    private List<String> columnType;
+
+    private static final Dictionary trieRoot = new Dictionary();
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: what is trieRoot ?

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+        SelectParameter.SelectParameterBuilder builder = selectParameter(locale)
+                .item(COUNTRY_NAME, COUNTRY_NAME)
+                .item(COUNTRY_CODE_ISO2, COUNTRY_CODE_ISO2)
+                .item(COUNTRY_CODE_ISO3, COUNTRY_CODE_ISO3)
+                .item(COUNTRY_NUMBER, COUNTRY_NUMBER)
+                .canBeBlank(false)
+                .name(FROM_UNIT_PARAMETER);
+
+        if (columnType != null) {
+            if (columnType.contains(SemanticCategoryEnum.COUNTRY_CODE_ISO2.getId())) {
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: this part is not tested

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+                builder.defaultValue(COUNTRY_CODE_ISO2);
+            } else if (columnType.contains(SemanticCategoryEnum.COUNTRY_CODE_ISO3.getId())) {
+                builder.defaultValue(COUNTRY_CODE_ISO3);
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: same here

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+            } else {
+                builder.defaultValue(COUNTRY_NAME);
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: same here

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        SelectParameter.SelectParameterBuilder secondBuilder = selectParameter(locale)
+                .item(ENGLISH_COUNTRY_NAME, ENGLISH_COUNTRY_NAME)
+                .item(FRENCH_COUNTRY_NAME, FRENCH_COUNTRY_NAME)
+                .item(COUNTRY_CODE_ISO2, COUNTRY_CODE_ISO2)
+                .item(COUNTRY_CODE_ISO3, COUNTRY_CODE_ISO3)
+                .item(COUNTRY_NUMBER, COUNTRY_NUMBER)
+                .canBeBlank(false)
+                .name(TO_UNIT_PARAMETER);
+        if (columnType != null) {
+            if (columnType.contains(SemanticCategoryEnum.COUNTRY_CODE_ISO2.getId())
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: not tested also

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+                    || columnType.contains(SemanticCategoryEnum.COUNTRY_CODE_ISO3.getId())) {
+                secondBuilder.defaultValue(ENGLISH_COUNTRY_NAME);
+            } else {
+                secondBuilder.defaultValue(COUNTRY_CODE_ISO2);
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: same here

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+
+    @Override
+    public void applyOnColumn(DataSetRow row, ActionContext context) {
+        final String columnId = context.getColumnId();
+        final String columnValue = row.get(columnId);
+        final String targetColumnId = ActionsUtils.getTargetColumnId(context);
+        int countryId = -1;
+        CountryCode result;
+
+        switch (context.getParameters().get(FROM_UNIT_PARAMETER)) {
```
[the remark this answers is not in the document]
Comment (tlnd-jbrun) [reply dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: could be good good to add a test also on default case

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+                countryId = CountryCode.getByAlpha2Code(columnValue).getNumeric();
+            }
+            break;
+        case COUNTRY_CODE_ISO3:
+            if (CountryCode.getByAlpha3Code(columnValue) != null) {
+                countryId = CountryCode.getByAlpha3Code(columnValue).getNumeric();
+            }
+            break;
+        case COUNTRY_NUMBER:
+            if (NumericHelper.isBigDecimal(columnValue)) {
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: not tested

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+                countryId = BigDecimalParser.toBigDecimal(columnValue).intValue();
+            }
+            break;
+        }
+
+        if (countryId != -1) {
+            result = CountryCode.getByCode(countryId);
+            switch (context.getParameters().get(TO_UNIT_PARAMETER)) {
+            case ENGLISH_COUNTRY_NAME:
+                row.set(targetColumnId, result.getName());
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: not tested

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+                break;
+            case FRENCH_COUNTRY_NAME:
+                Locale tempLocale = new Locale("", result.getAlpha2());
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: not tested

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+        /**
+         * Search through the dictionary for a word.
+         * 
+         * @param string The word to search for.
+         * @return Whether or not the word exists in the dictionary.
+         */
+        public int search(String string) {
+            if (roots.containsKey(string.charAt(0))) {
+                if (string.length() == 1) {
+                    return roots.get(string.charAt(0)).value;
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: case not covered by test

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+                }
+                return searchFor(string.substring(1), roots.get(string.charAt(0)));
+            } else {
+                return -1;
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: case not covered by test

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+        }
+    }
+
+    static class Node {
+
+        protected Node parent;
+
+        protected int value = -1;
+
+        protected HashMap<Character, Node> children = new HashMap<>();
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: protected Map<Character, Node> children = new HashMap<>(); is enough ?

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+        for (SemanticDomain semanticDomain : column.getSemanticDomains()) {
+            semanticIds.add(semanticDomain.getId());
+        }
+        return new CountryConverter(semanticIds);
+    }
+
+    // Trie tree
+    static class Dictionary {
+
+        private HashMap<Character, Node> roots = new HashMap<>();
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: private Map<Character, Node> roots = new HashMap<>();
could be enough ?

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+                return searchFor(string.substring(1), roots.get(string.charAt(0)));
+            } else {
+                return -1;
+            }
+        }
+
+        /**
+         * Insert a word into the dictionary.
+         * 
+         * @param string The word to insert.
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: could you precise what is the value parameter ?

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+    @Override
+    public ActionDefinition adapt(ColumnMetadata column) {
+        List<String> semanticIds = new ArrayList<>();
+        for (SemanticDomain semanticDomain : column.getSemanticDomains()) {
+            semanticIds.add(semanticDomain.getId());
+        }
+        return new CountryConverter(semanticIds);
+    }
+
+    // Trie tree
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: I think you use this code ?
https://gist.github.com/prettymuchbryce/3719435/868ae3231e98b3bde667a9e899157e20dd58c366
 could you cite the reference please ?

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+    // Initialise trie tree only at the first use of the action.
+    static {
+        for (CountryCode countryCode : CountryCode.values()) {
+            if (countryCode.getAssignment().equals(CountryCode.Assignment.OFFICIALLY_ASSIGNED)) {
+                trieRoot.insert(countryCode.getName(), countryCode.getNumeric());
+            }
+        }
+        for (String countryCode : Locale.getISOCountries()) {
+            Locale obj = new Locale("", countryCode);
+            String countryName = obj.getDisplayCountry(Locale.FRENCH);
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: Why the Locale is French here ?

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+    public void applyOnColumn(DataSetRow row, ActionContext context) {
+        final String columnId = context.getColumnId();
+        final String columnValue = row.get(columnId);
+        final String targetColumnId = ActionsUtils.getTargetColumnId(context);
+        int countryId = -1;
+        CountryCode result;
+
+        switch (context.getParameters().get(FROM_UNIT_PARAMETER)) {
+        case COUNTRY_NAME:
+            countryId = trieRoot.search(columnValue);
```
Comment (tlnd-jbrun) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: why create a dictionary and do not use CountryCode.findByName(pattern) here ?

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+
+    public CountryConverter() {
+        // nothing to do here
+    }
+
+    private CountryConverter(List<String> columnType) {
+        this.columnType = columnType;
+    }
+
+    // Initialise trie tree only at the first use of the action.
```
[the remark this answers is not in the document]
Comment (tlnd-cgall) [reply dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: https://fr.wikipedia.org/wiki/Trie_(informatique)

File: dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+                countryId = CountryCode.getByAlpha3Code(columnValue).getNumeric();
+            }
+            break;
+        case COUNTRY_NUMBER:
+            if (NumericHelper.isBigDecimal(columnValue)) {
+                countryId = BigDecimalParser.toBigDecimal(columnValue).intValue();
+            }
+            break;
+        default:
+            // theoretically impossible case
```
Comment (yacarof) [inline dataprep-actions/src/main/java/org/talend/dataprep/transformation/actions/conversions/CountryConverter.java]: As this code could be reached by a direct API call I would at least log a warning message, and better fail with an exception.

**Pull request merged**